File: db4oj/core/src/com/db4o/cluster/inside/ClusterConstraint.java
/* Copyright (C) 2004 - 2005  db4objects Inc.  http://www.db4o.com */

package com.db4o.cluster.inside;

import com.db4o.cluster.*;
import com.db4o.inside.*;
import com.db4o.query.*;


public class ClusterConstraint implements Constraint{
    
    private final Cluster _cluster;
    final Constraint[] _constraints;
    
    public ClusterConstraint(Cluster cluster, Constraint[] constraints){
        _cluster = cluster; 
        _constraints = constraints;
    }
    
    private ClusterConstraint compatible (Constraint with){
        if(! (with instanceof ClusterConstraint)){
            throw new IllegalArgumentException();
        }
        ClusterConstraint other = (ClusterConstraint)with;
        if(other._constraints.length != _constraints.length){
            throw new IllegalArgumentException();
        }
        return other;
    }

    public Constraint and(Constraint with) {
        return join(with, true);
    }

    public Constraint or(Constraint with) {
        return join(with, false);
    }
    
    private Constraint join(Constraint with, boolean isAnd){
        ClusterConstraint other = compatible(with);
        Constraint[] newConstraints = new Constraint[_constraints.length];
        for (int i = 0; i < _constraints.length; i++) {
            newConstraints[i] = isAnd ? _constraints[i].and(other._constraints[i]) : _constraints[i].or(other._constraints[i]);
        }
        return new ClusterConstraint(_cluster, newConstraints);
    }
    

    public Constraint equal() {
        for (int i = 0; i < _constraints.length; i++) {
            _constraints[i].equal();
        }
        return this;
    }

    public Constraint greater() {
        for (int i = 0; i < _constraints.length; i++) {
            _constraints[i].greater();
        }
        return this;
    }

    public Constraint smaller() {
        for (int i = 0; i < _constraints.length; i++) {
            _constraints[i].smaller();
        }
        return this;
    }

    public Constraint identity() {
        for (int i = 0; i < _constraints.length; i++) {
            _constraints[i].identity();
        }
        return this;
    }

    public Constraint like() {
        for (int i = 0; i < _constraints.length; i++) {
            _constraints[i].like();
        }
        return this;
    }

    public Constraint contains() {
        for (int i = 0; i < _constraints.length; i++) {
            _constraints[i].contains();
        }
        return this;
    }

    public Constraint not() {
        for (int i = 0; i < _constraints.length; i++) {
            _constraints[i].not();
        }
        return this;
    }

    public Object getObject() {
        Exceptions4.notSupported();
        return null;
    }

}

File: db4on/core/src/core/com/db4o/inside/cluster/ClusterConstraint.cs
namespace com.db4o.inside.cluster
{
	/// <exclude></exclude>
	public class ClusterConstraint : com.db4o.query.Constraint
	{
		internal readonly com.db4o.cluster.Cluster _cluster;

		internal readonly com.db4o.query.Constraint[] _constraints;

		public ClusterConstraint(com.db4o.cluster.Cluster cluster, com.db4o.query.Constraint[]
			 constraints)
		{
			_cluster = cluster;
			_constraints = constraints;
		}

		private com.db4o.inside.cluster.ClusterConstraint compatible(com.db4o.query.Constraint
			 with)
		{
			if (!(with is com.db4o.inside.cluster.ClusterConstraint))
			{
				throw new System.ArgumentException();
			}
			com.db4o.inside.cluster.ClusterConstraint other = (com.db4o.inside.cluster.ClusterConstraint
				)with;
			if (other._constraints.Length != _constraints.Length)
			{
				throw new System.ArgumentException();
			}
			return other;
		}

		public virtual com.db4o.query.Constraint and(com.db4o.query.Constraint with)
		{
			return join(with, true);
		}

		public virtual com.db4o.query.Constraint or(com.db4o.query.Constraint with)
		{
			return join(with, false);
		}

		private com.db4o.query.Constraint join(com.db4o.query.Constraint with, bool isAnd
			)
		{
			lock (_cluster)
			{
				com.db4o.inside.cluster.ClusterConstraint other = compatible(with);
				com.db4o.query.Constraint[] newConstraints = new com.db4o.query.Constraint[_constraints
					.Length];
				for (int i = 0; i < _constraints.Length; i++)
				{
					newConstraints[i] = isAnd ? _constraints[i].and(other._constraints[i]) : _constraints
						[i].or(other._constraints[i]);
				}
				return new com.db4o.inside.cluster.ClusterConstraint(_cluster, newConstraints);
			}
		}

		public virtual com.db4o.query.Constraint equal()
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].equal();
				}
				return this;
			}
		}

		public virtual com.db4o.query.Constraint greater()
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].greater();
				}
				return this;
			}
		}

		public virtual com.db4o.query.Constraint smaller()
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].smaller();
				}
				return this;
			}
		}

		public virtual com.db4o.query.Constraint identity()
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].identity();
				}
				return this;
			}
		}

		public virtual com.db4o.query.Constraint like()
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].like();
				}
				return this;
			}
		}

		public virtual com.db4o.query.Constraint startsWith(bool caseSensitive)
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].startsWith(caseSensitive);
				}
				return this;
			}
		}

		public virtual com.db4o.query.Constraint endsWith(bool caseSensitive)
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].endsWith(caseSensitive);
				}
				return this;
			}
		}

		public virtual com.db4o.query.Constraint contains()
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].contains();
				}
				return this;
			}
		}

		public virtual com.db4o.query.Constraint not()
		{
			lock (_cluster)
			{
				for (int i = 0; i < _constraints.Length; i++)
				{
					_constraints[i].not();
				}
				return this;
			}
		}

		public virtual object getObject()
		{
			com.db4o.inside.Exceptions4.notSupported();
			return null;
		}
	}
}
